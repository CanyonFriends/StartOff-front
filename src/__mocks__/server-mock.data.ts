import { ProfileServerType, ProjectServerResponseType, SkillServerType } from '../@types/server';
import generateUUID from '../utils/generateUUID';

export const makeSkillMock = ({
  skill_id = generateUUID(),
  skill_name = 'typescript',
  color = '#eeeeee',
  text_color = '#000000',
}): SkillServerType => {
  return {
    skill_id,
    skill_name,
    color,
    text_color,
  };
};

export const makeProjectResponseMock = ({
  id = Math.floor(Math.random() * 10000),
  title = 'title',
  content = 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s,',
  deploy_url = 'https://shellboylog.com/list',
  github_url = 'https://github.com/qkrdmstlr3/devlog',
  introduce = 'introduce introduce introduce',
  start_date = '1998-01-22',
  end_date = '2021-01-22',
  project_skills = [makeSkillMock({ skill_name: 'typescript' }), makeSkillMock({ skill_name: 'javascript' })],
}): ProjectServerResponseType => {
  return {
    id,
    title,
    content,
    deploy_url,
    github_url,
    introduce,
    start_date,
    end_date,
    project_skills,
  };
};

export const makeProfileMock = ({
  baekjoon_id = 'shellboy',
  blog_url = 'https://shellboylog.com',
  github_url = 'https://github.com/qkrdmstlr3',
  introduce = 'hello my name is shellboy',
  nickname = 'shellboy',
  projects = [makeProjectResponseMock({}), makeProjectResponseMock({})],
  user_skills = [makeSkillMock({ skill_name: 'typescript' }), makeSkillMock({ skill_name: 'javascript' })],
}): ProfileServerType => {
  return {
    baekjoon_id,
    blog_url,
    github_url,
    introduce,
    nickname,
    projects,
    user_skills,
  };
};
