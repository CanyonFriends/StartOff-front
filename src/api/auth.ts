import axios from 'axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  uuid: string;
  email: string;
  nickname: string;
  user_id: string;
}

export interface SignupResponseType {
  user_id: string;
}

export interface LogoutRequestType {
  email: string;
  accessToken: string;
  uuid: string;
}

export interface LogoutResponseType {
  msg: string;
  success: boolean;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType): Promise<ErrorType | SignupResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response.data as SignupResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const logoutAPI = async ({
  email,
  accessToken,
  uuid,
}: LogoutRequestType): Promise<ErrorType | LogoutResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/logout',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      data: {
        email,
        accessToken,
        uuid,
      },
    });
    return response.data as LogoutResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};
