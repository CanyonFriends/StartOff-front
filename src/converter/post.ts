import { BoardServerType, SummarizedPostServerType, CreatePostServerType, PostServerType } from '../@types/server';
import { BoardClientType, SummarizedPostClientType, CreatePostClientType, PostClientType } from '../@types/client';
import { skillServerType2ClientType } from './skill';

export const summarizedPostServerType2ClientType = (
  summarizedPost: SummarizedPostServerType,
): SummarizedPostClientType => {
  return {
    postId: String(summarizedPost.post_id),
    title: summarizedPost.title,
    currentPeople: summarizedPost.current_people,
    maxPeople: summarizedPost.max_people,
    createAt: new Date(summarizedPost.created_at),
    nickname: summarizedPost.nickname,
    postSkills: summarizedPost.post_skills.map((skill) => skillServerType2ClientType(skill)),
  };
};

export const boardServerType2ClientType = (board: BoardServerType): BoardClientType => {
  return {
    content: board.content.map((post) => summarizedPostServerType2ClientType(post)),
    totalElements: board.totalElements,
    totalPages: board.totalPages,
  };
};

export const createPostClientType2ServerType = (post: CreatePostClientType): CreatePostServerType => {
  return {
    category: post.category.toUpperCase(),
    content: post.content,
    current_people: post.currentPeople,
    max_people: post.maxPeople,
    post_skills: post.postSkills.map((skill) => skill.skillName),
    title: post.title,
    user_id: Number(post.userId),
  };
};

export const postServerType2ClientType = (post: PostServerType): PostClientType => {
  return {
    postId: String(post.post_id),
    category: post.category,
    title: post.title,
    content: post.content,
    createdAt: new Date(post.created_at),
    currentPeople: post.current_people,
    maxPeople: post.max_people,
    nickname: post.nickname,
    postSkills: post.post_skills.map((skill) => skillServerType2ClientType(skill)),
    userId: String(post.user_id),
  };
};
