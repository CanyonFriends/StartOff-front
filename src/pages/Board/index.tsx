import React from 'react';
import BoardTemplate from './template';

function Board() {
  return <BoardTemplate posts={[]} />;
}

export default Board;
