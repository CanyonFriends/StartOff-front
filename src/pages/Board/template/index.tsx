import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { Link, useHistory } from 'react-router-dom';
import * as Style from './styled';
import { SummarizedPostClientType } from '../../../@types/client';
import CommonHeader from '../../../components/Layout/CommonHeader';
import { Button } from '../../../components/UI/atom';
import { Dropdown, SearchBox } from '../../../components/UI/molecule';
import { AlertModal, Pagination, SummarizedPostItem } from '../../../components/UI/organism';
import { RootState } from '../../../redux/store';
import { UserState } from '../../../redux/user/types';
import { buildCreatePostPath, buildPostPath } from '../../../Routes';

interface BoardTemplateProps {
  posts: SummarizedPostClientType[];
  board: string;
  totalPage: number;
  currentPage: number;
  getPosts: (query: string, type?: string[]) => Promise<void>;
  handlePagination: (pageNumber: number) => void;
}

const DROPDOWN_ITEM: { [name: string]: string } = {
  제목: 'title',
  내용: 'content',
  태그: 'skill',
};
function BoardTemplate({ posts, board, totalPage, currentPage, handlePagination, getPosts }: BoardTemplateProps) {
  const history = useHistory();
  const [error, setError] = useState('');
  const [dropdownItems, setDropdownItems] = useState(
    Object.keys(DROPDOWN_ITEM).map((item) => ({ text: item, id: item, selected: false })),
  );
  const { isSignin } = useSelector<RootState>((state) => state.user) as UserState;

  const clickDropdownItem = (id: string) => {
    const updatedSelectedDropdownItems = dropdownItems.map((item) => {
      if (item.id === id) item.selected = !item.selected;
      return item;
    });
    setDropdownItems(updatedSelectedDropdownItems);
  };

  const searchSubmit = async (query: string) => {
    const selectedItems = dropdownItems.filter((item) => item.selected).map((item) => DROPDOWN_ITEM[item.text]);
    await getPosts(query, selectedItems);

    return '';
  };

  const clickCreatePostButton = () => {
    if (!isSignin) {
      setError('로그인이 필요합니다');
      return;
    }

    history.push(buildCreatePostPath(board));
  };

  const closeAlertModal = () => {
    setError('');
  };

  return (
    <>
      {!!error && <AlertModal content={error} clickCloseButton={closeAlertModal} />}
      <CommonHeader />
      <Style.Container>
        <Style.Header>
          <Style.HeaderLeft>
            <Dropdown
              placeholder={
                dropdownItems
                  .filter((item) => item.selected)
                  .map((item) => item.text)
                  .join('. ') || '카테고리 선택'
              }
              items={dropdownItems}
              clickItem={clickDropdownItem}
            />
            <SearchBox searchSubmit={searchSubmit} />
          </Style.HeaderLeft>
          <Button theme="secondary" size="medium" onClick={clickCreatePostButton}>
            글 쓰기
          </Button>
        </Style.Header>
        <Style.PostList>
          {posts.map((post) => (
            <Link to={buildPostPath(board, post.postId)} key={post.postId}>
              <SummarizedPostItem post={post} />
            </Link>
          ))}
        </Style.PostList>
        <Pagination totalCount={totalPage} currentPage={currentPage} onClickPageButton={handlePagination} />
      </Style.Container>
    </>
  );
}

export default BoardTemplate;
