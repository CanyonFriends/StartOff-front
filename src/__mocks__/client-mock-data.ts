import { ErrorType } from '../api/error';
import { SigninResponseType } from '../api/auth';
import {
  SkillClientType,
  ProjectClientType,
  ProfileClientType,
  SummarizedPostClientType,
  BoardClientType,
  CreatePostClientType,
  PostClientType,
} from '../@types/client';
import generateUUID from '../utils/generateUUID';

// api
export const makeSigninSuccessInfoMock = ({
  access_token = 'access_token',
  uuid = 'uuid',
  email = 'email@email.com',
  nickname = 'nickname',
  user_id = '1',
}): SigninResponseType => {
  return {
    access_token,
    uuid,
    email,
    nickname,
    user_id,
  };
};

// error
export const makeErrorMock = ({ error_msg = '에러 발생' }): ErrorType => {
  return { error_msg };
};

// data
export const makeSkillMock = ({
  skillId = generateUUID(),
  skillName = 'c#',
  color = '#eeeeee',
  textColor = '#000000',
}): SkillClientType => {
  return {
    skillId,
    skillName,
    color,
    textColor,
  };
};

export const makeProjectMock = ({
  id = Math.floor(Math.random() * 10000),
  title = 'project',
  deployUrl = 'https://shellboylog.com/list',
  githubUrl = 'https://github.com/qkrdmstlr3/devlog',
  introduce = 'introduce introduce introduce',
  startDate = new Date('1998-01-22'),
  endDate = new Date('2021-01-22'),
  content = 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s,',
  projectSklls = [makeSkillMock({ skillName: 'typescript' }), makeSkillMock({ skillName: 'c#' })],
}): ProjectClientType => {
  return {
    id,
    title,
    deployUrl,
    githubUrl,
    introduce,
    startDate,
    endDate,
    content,
    projectSklls,
  };
};

export const makeProfileMock = ({
  baekjoonId = 'shellboy',
  blogUrl = 'https://shellboylog.com',
  githubUrl = 'https://github.com/qkrdmstlr3',
  introduce = 'hello my name is shellboy',
  nickname = 'shellboy',
  projects = [makeProjectMock({})],
  userSkills = [makeSkillMock({ skillName: 'typescript' }), makeSkillMock({ skillName: 'c#' })],
}): ProfileClientType => {
  return {
    baekjoonId,
    blogUrl,
    githubUrl,
    introduce,
    nickname,
    projects,
    userSkills,
  };
};

export const makeMockSummarizedPost = ({
  postId = generateUUID(),
  title = 'title',
  currentPeople = 5,
  maxPeople = 10,
  createAt = new Date(),
  nickname = 'user',
  postSkills = [makeSkillMock({ skillName: 'typescript' }), makeSkillMock({ skillName: 'c#' })],
}): SummarizedPostClientType => {
  return {
    postId,
    title,
    currentPeople,
    maxPeople,
    createAt,
    postSkills,
    nickname,
  };
};

export const makeMockBoard = ({
  content = [makeMockSummarizedPost({})],
  totalElements = 1,
  totalPages = 1,
}): BoardClientType => {
  return {
    content,
    totalElements,
    totalPages,
  };
};

export const makeCreatePost = ({
  category = 'category',
  content = 'content',
  currentPeople = 5,
  maxPeople = 10,
  postSkills = [makeSkillMock({})],
  title = 'title',
  userId = '1',
}): CreatePostClientType => {
  return {
    category,
    content,
    currentPeople,
    maxPeople,
    postSkills,
    title,
    userId,
  };
};

export const makePostMock = ({
  postId = generateUUID(),
  category = 'category',
  title = 'title',
  content = 'content',
  createdAt = new Date('1998-01-22'),
  currentPeople = 5,
  maxPeople = 10,
  nickname = 'shellboy',
  postSkills = [makeSkillMock({})],
}): PostClientType => {
  return {
    postId,
    category,
    title,
    content,
    createdAt,
    currentPeople,
    maxPeople,
    nickname,
    postSkills,
  };
};

export const mockString =
  "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";
