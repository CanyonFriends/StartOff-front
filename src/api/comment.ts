import { AxiosError } from 'axios';
import { ErrorType } from './error';
import axios from '../utils/axios';
import { CommentServerType } from '../@types/server';
import { commentServerType2ClientType } from '../converter/comment';

export interface CreateCommentRequest {
  postId: string;
  userId: string;
  content: string;
}

export interface UpdateCommentRequest {
  postId: string;
  userId: string;
  content: string;
  commentId: string;
}

export interface DeleteCommentRequest {
  postId: string;
  commentId: string;
}

export const createCommentAPI = async ({ postId, userId, content }: CreateCommentRequest) => {
  try {
    const response = await axios({
      method: 'POST',
      url: `/v1/posts/${postId}/comments`,
      data: {
        parend_id: null,
        user_id: userId,
        content,
      },
    });

    return commentServerType2ClientType(response.data as CommentServerType);
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const updateCommentAPI = async ({ postId, userId, content, commentId }: UpdateCommentRequest) => {
  try {
    const response = await axios({
      method: 'PUT',
      url: `/v1/posts/${postId}/comments/${commentId}`,
      data: {
        parend_id: null,
        user_id: userId,
        content,
      },
    });

    return commentServerType2ClientType(response.data as CommentServerType);
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const deleteCommentAPI = async ({ postId, commentId }: DeleteCommentRequest) => {
  try {
    await axios({
      method: 'DELETE',
      url: `/v1/posts/${postId}/comments/${commentId}`,
    });

    return true;
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};
