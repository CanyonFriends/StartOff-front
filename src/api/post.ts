import axios from '../utils/axios';
import { ErrorType } from './error';
import { BoardServerType } from '../@types/server';
import { boardServerType2ClientType, createPostClientType2ServerType } from '../converter/post';
import { CreatePostClientType } from '../@types/client';

export const getCategoriesAPI = async () => {
  try {
    const response = await axios({
      method: 'GET',
      url: '/v1/categories',
    });

    return response.data as string[];
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const getPostsAPI = async (page: number, size: number, category: string) => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/posts?page=${page}&size=${size}&category=${category}`,
    });

    return boardServerType2ClientType(response.data as BoardServerType);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const createPostAPI = async (post: CreatePostClientType) => {
  try {
    await axios({
      method: 'POST',
      url: '/v1/posts',
      data: {
        ...createPostClientType2ServerType(post),
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
