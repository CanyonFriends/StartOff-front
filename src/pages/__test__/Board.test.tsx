/**
 * @jest-environment jsdom
 */
import React from 'react';
import { waitFor, fireEvent } from '@testing-library/react';
import { render } from '../../test-utils';
import Board from '../Board';
import { getCategoriesAPI, getPostsAPI } from '../../api/post';
import { makeMockBoard } from '../../__mocks__/client-mock-data';

jest.mock('../../api/post');

const getPostsMockAPI = getPostsAPI as jest.MockedFunction<typeof getPostsAPI>;
const getCategoriesMockAPI = getCategoriesAPI as jest.MockedFunction<typeof getCategoriesAPI>;

beforeEach(() => {
  jest.resetAllMocks();
  getPostsMockAPI.mockReturnValue(new Promise((res) => res(makeMockBoard({}))));
  getCategoriesMockAPI.mockReturnValue(new Promise((res) => res([])));
});

describe('<Board> 페이지', () => {
  it('snapshot 체크', async () => {
    const component = render(<Board />);
    await waitFor(() => expect(component.container).toMatchSnapshot());
  });

  it('렌더링 테스트', async () => {
    const component = render(<Board />);

    await waitFor(() => {
      component.getByText('글 쓰기');
      component.getByText('검색');
    });
  });

  it('로그인 없이 글 쓰기 버튼 클릭 시 에러', async () => {
    const component = render(<Board />);

    await waitFor(() => {
      const createPostButton = component.getByText('글 쓰기');
      fireEvent.click(createPostButton);
    });

    await component.findByText('로그인이 필요합니다');
  });
});
