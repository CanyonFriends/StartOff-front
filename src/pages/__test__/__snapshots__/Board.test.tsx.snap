// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board> 페이지 snapshot 체크 1`] = `
<div>
  <div
    id="modal"
  />
  게시글을 불러오는데 실패하였습니다
</div>
`;
